File: evaluation/S0/java/Hard/2688-count-the-number-of-infection-sequences.java
// Source: https://leetcode.com/problems/count-the-number-of-infection-sequences/   |   Difficulty: Hard
//
// Problem Description:
// You are given an integer n and an array sick sorted in increasing order, representing positions of infected people in a line of n people.
//
// At each step, one uninfected person adjacent to an infected person gets infected. This process continues until everyone is infected.
//
// An infection sequence is the order in which uninfected people become infected, excluding those initially infected.
//
// Return the number of different infection sequences possible, modulo 109+7.
//
// Example:
// Input: n = 5, sick = [0,4]
//
// Output: 4
//
// Explanation:
//
// There is a total of 6 different sequences overall.
//
//
// 	Valid infection sequences are [1,2,3], [1,3,2], [3,2,1] and [3,1,2].
// 	[2,3,1] and [2,1,3] are not valid infection sequences because the person at index 2 cannot be infected at the first step.
//
// Constraints:
// 2 <= n <= 105
// 	1 <= sick.length <= n - 1
// 	0 <= sick[i] <= n - 1
// 	sick is sorted in increasing order.
//

class Solution {
    public int numberOfSequence(int n, int[] sick) {
        final int MOD = 1_000_000_007;

        // Precompute factorials and their modular inverses
        long[] fact = new long[n + 1];
        long[] invFact = new long[n + 1];
        fact[0] = 1;
        for (int i = 1; i <= n; i++) {
            fact[i] = fact[i - 1] * i % MOD;
        }
        invFact[n] = modInverse(fact[n], MOD);
        for (int i = n - 1; i >= 0; i--) {
            invFact[i] = invFact[i + 1] * (i + 1) % MOD;
        }

        long result = 1;
        int prevSick = -1;

        for (int sickPos : sick) {
            int gap = sickPos - prevSick - 1;
            if (gap > 0) {
                result = result * fact[gap] % MOD;
            }
            prevSick = sickPos;
        }

        int lastGap = n - prevSick - 1;
        if (lastGap > 0) {
            result = result * fact[lastGap] % MOD;
        }

        return (int) result;            
    }
}

File: evaluation/S0/javascript/Hard/2688-count-the-number-of-infection-sequences.js
// Source: https://leetcode.com/problems/count-the-number-of-infection-sequences/   |   Difficulty: Hard
//
// Problem Description:
// You are given an integer n and an array sick sorted in increasing order, representing positions of infected people in a line of n people.
//
// At each step, one uninfected person adjacent to an infected person gets infected. This process continues until everyone is infected.
//
// An infection sequence is the order in which uninfected people become infected, excluding those initially infected.
//
// Return the number of different infection sequences possible, modulo 109+7.
//
// Example:
// Input: n = 5, sick = [0,4]
//
// Output: 4
//
// Explanation:
//
// There is a total of 6 different sequences overall.
//
//
// 	Valid infection sequences are [1,2,3], [1,3,2], [3,2,1] and [3,1,2].
// 	[2,3,1] and [2,1,3] are not valid infection sequences because the person at index 2 cannot be infected at the first step.
//
// Constraints:
// 2 <= n <= 105
// 	1 <= sick.length <= n - 1
// 	0 <= sick[i] <= n - 1
// 	sick is sorted in increasing order.
//

/**
 * @param {number} n
 * @param {number[]} sick
 * @return {number}
 */
var numberOfSequence = function(n, sick) {
    const mod = 1e9 + 7;
    const m = sick.length;
    const dp = Array(m + 1).fill(0);
    dp[0] = 1;
    for (let i = 1; i <= m; i++) {
        dp[i] = (dp[i - 1] * i) % mod;
    }
    let result = 1;
    for (let i = 1; i < m; i++) {
        const gap = sick[i] - sick[i - 1] - 1;
        if (gap > 0) {
            result = (result * dp[gap]) % mod;
        }
    }
    const leftGap = sick[0];
    const rightGap = n - 1 - sick[m - 1];
    if (leftGap > 0) {
        result = (result * dp[leftGap]) % mod;
    }
    if (rightGap > 0) {
        result = (result * dp[rightGap]) % mod;
    }
    return result;  
};
